src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseBaseKernels.hpp

```diff
@@ -494,6 +494,269 @@ struct SolidInternalEnergyUpdateKernel
+ */
+class ScalingForSystemSolutionKernel : public isothermalCompositionalMultiphaseBaseKernels::ScalingForSystemSolutionKernel
+{
+public:
+
+  using Base = isothermalCompositionalMultiphaseBaseKernels::ScalingForSystemSolutionKernel;
+  using Base::m_numComp;
+  using Base::m_localSolution;
+
+  /**
+   * @brief Create a new kernel instance
+   * @param[in] maxRelativePresChange the max allowed relative pressure change
+   * @param[in] maxRelativeTempChange the max allowed relative temperature change
+   * @param[in] maxCompFracChange the max allowed comp fraction change
+   * @param[in] rankOffset the rank offset
+   * @param[in] numComp the number of components
+   * @param[in] dofKey the dof key to get dof numbers
+   * @param[in] subRegion the subRegion
+   * @param[in] localSolution the Newton update
+   * @param[in] pressure the pressure vector
+   * @param[in] temperature the temperature vector
+   * @param[in] compDens the component density vector
+   */
+  ScalingForSystemSolutionKernel( real64 const maxRelativePresChange,
+                                  real64 const maxRelativeTempChange,
+                                  real64 const maxCompFracChange,
+                                  globalIndex const rankOffset,
+                                  integer const numComp,
+                                  string const dofKey,
+                                  ElementSubRegionBase const & subRegion,
+                                  arrayView1d< real64 const > const localSolution,
+                                  arrayView1d< real64 const > const pressure,
+                                  arrayView1d< real64 const > const temperature,
+                                  arrayView2d< real64 const, compflow::USD_COMP > const compDens )
+    : Base( maxRelativePresChange,
+            maxCompFracChange,
+            rankOffset,
+            numComp,
+            dofKey,
+            subRegion,
+            localSolution,
+            pressure,
+            compDens ),
+    m_maxRelativeTempChange( maxRelativeTempChange ),
+    m_temperature( temperature )
+  {}
+
+  /**
+   * @brief Compute the local value of the scaling factor
+   * @tparam FUNC the type of the function that can be used to customize the kernel
+   * @param[in] ei the element index
+   * @param[inout] stack the stack variables
+   * @param[in] kernelOp the function used to customize the kernel
+   */
+  GEOSX_HOST_DEVICE
+  void computeScalingFactor( localIndex const ei,
+                             StackVariables & stack ) const
+  {
+    real64 constexpr eps =
+      isothermalCompositionalMultiphaseBaseKernels::minDensForDivision;
+
+    Base::computeScalingFactor( ei, stack, [&] ()
+    {
+      // compute the change in temperature
+      real64 const temp = m_temperature[ei];
+      if( temp > eps )
+      {
+        real64 const absTempChange = LvArray::math::abs( m_localSolution[stack.localRow + m_numComp + 1] );
```

Comment:
As we repeat the scaling operation quite a few time over the classes, is there a solution in which it can be factorized and in which `m_maxRelativeXChange`, `local_offfset` and `m_X` are picked from argument ?

Reply:
Yes, we could have a utility function in the kernel class that does that for us to avoid some redundancy in the code. I have not implemented it yet, but I am going to think about a good way